### ISSUE TITLE
chef_solo provisioning path being changed between `vagrant up` and `vagrant provision`

### ISSUE THREAD

**Issue by tknerr** [mitchellh/vagrant#4988]
...which leads to this error message:
```
Shared folders that Chef requires are missing on the virtual machine.
This is usually due to configuration changing after already booting the
machine. The fix is to run a `vagrant reload` so that the proper shared
folders will be prepared and mounted on the VM. 
```

# Steps to reproduce

 1. create a simple Vagrantfile with `:chef_solo` provisioner
 2. run `vagrant up`
 3. check the created provisioner paths via `vagrant ssh -c "ls -la /tmp/vagrant-chef-*/" 
    => is 
 4. run `vagrant provision`

# Example

I'm using this Vagrantfile:
```ruby
Vagrant::configure("2") do |config|

  # the Chef version to use
  config.omnibus.chef_version = "11.16.4"
  
  # common basebox for all VMs
  config.vm.box = "chef/ubuntu-12.04-i386"

  config.vm.define :foo do |foo_config|
    foo_config.vm.provision :chef_solo do |chef|
      chef.cookbooks_path = "./cookbooks"
      chef.add_recipe "hello"
    end
  end
end
```

Next, running `vagrant up` shows that "/tmp/vagrant-chef-3/" is created:
```
D:\Repos\_github\bills-kitchen\target\build\repo\tmp>vagrant up
Bringing machine 'foo' up with 'virtualbox' provider...
==> foo: Importing base box 'chef/ubuntu-12.04-i386'...
==> foo: Matching MAC address for NAT networking...
==> foo: Setting the name of the VM: tmp_foo_1418711517976_72019
==> foo: Clearing any previously set network interfaces...
==> foo: Preparing network interfaces based on configuration...
    foo: Adapter 1: nat
==> foo: Forwarding ports...
    foo: 22 => 2222 (adapter 1)
==> foo: Booting VM...
==> foo: Waiting for machine to boot. This may take a few minutes...
    foo: SSH address: 127.0.0.1:2222
    foo: SSH username: vagrant
    foo: SSH auth method: private key
    foo: Warning: Connection timeout. Retrying...
    foo:
    foo: Vagrant insecure key detected. Vagrant will automatically replace
    foo: this with a newly generated keypair for better security.
    foo:
    foo: Inserting generated public key within guest...
    foo: Removing insecure key from the guest if its present...
    foo: Key inserted! Disconnecting and reconnecting using new SSH key...
==> foo: Machine booted and ready!
==> foo: Checking for guest additions in VM...
==> foo: Mounting shared folders...
    foo: /vagrant => D:/Repos/_github/bills-kitchen/target/build/repo/tmp
    foo: /tmp/vagrant-cache => D:/Repos/_github/bills-kitchen/target/build/home/.vagrant.d/cache/chef/ubuntu-12.04-i386
    foo: /tmp/vagrant-chef-3/chef-solo-1/cookbooks => D:/Repos/_github/bills-kitchen/target/build/repo/tmp/cookbooks
==> foo: Installing Chef 11.16.4 Omnibus package...
==> foo: Downloading Chef 11.16.4 for ubuntu...
==> foo: downloading https://www.chef.io/chef/metadata?v=11.16.4&prerelease=false&nightlies=false&p=ubuntu&pv=12.04&m=x86_64
==> foo:   to file /tmp/install.sh.1161/metadata.txt
==> foo: trying wget...
==> foo: url    https://opscode-omnibus-packages.s3.amazonaws.com/ubuntu/12.04/x86_64/chef_11.16.4-1_amd64.deb
==> foo: md5    2ffff5b4d80e4dcffc917f8eb2003a31
==> foo: sha256 28b08975e7e33ac46c888616ec7fa232a0c624aeeda81e58a6047d2c6b62edfb
==> foo: downloaded metadata file looks valid...
==> foo: /tmp/vagrant-cache/vagrant_omnibus/chef_11.16.4-1_amd64.deb already exists, verifiying checksum...
==> foo: Comparing checksum with sha256sum...
==> foo: checksum compare succeeded, using existing file!
==> foo: Installing Chef 11.16.4
==> foo: installing with dpkg...
==> foo: Selecting previously unselected package chef.
==> foo: (Reading database ...
==> foo: 56035 files and directories currently installed.)
==> foo: Unpacking chef (from .../chef_11.16.4-1_amd64.deb) ...
==> foo: Setting up chef (11.16.4-1) ...
==> foo: Thank you for installing Chef!
==> foo: Configuring cache buckets...
==> foo: Running provisioner: chef_solo...
==> foo: Detected Chef (latest) is already installed
Generating chef JSON and uploading...
==> foo: Running chef-solo...
==> foo: stdin: is not a tty
==> foo: [2014-12-16T06:33:18+00:00] INFO: Forking chef instance to converge...
==> foo: [2014-12-16T06:33:18+00:00] WARN:
==> foo: * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
==> foo: SSL validation of HTTPS requests is disabled. HTTPS connections are still
==> foo: encrypted, but chef is not able to detect forged replies or man in the middle
==> foo: attacks.
==> foo:
==> foo: To fix this issue add an entry like this to your configuration file:
==> foo:
==> foo: ```
==> foo:   # Verify all HTTPS connections (recommended)
==> foo:   ssl_verify_mode :verify_peer
==> foo:
==> foo:   # OR, Verify only connections to chef-server
==> foo:   verify_api_cert true
==> foo: ```
==> foo:
==> foo: To check your SSL configuration, or troubleshoot errors, you can use the
==> foo: `knife ssl check` command like so:
==> foo:
==> foo: ```
==> foo:   knife ssl check -c /tmp/vagrant-chef-3/solo.rb
==> foo: ```
==> foo:
==> foo: * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
==> foo: [2014-12-16T06:33:18+00:00] INFO: *** Chef 11.16.4 ***
==> foo: [2014-12-16T06:33:18+00:00] INFO: Chef-client pid: 1441
==> foo: [2014-12-16T06:33:19+00:00] INFO: Setting the run_list to ["recipe[hello]"] from CLI options
==> foo: [2014-12-16T06:33:19+00:00] INFO: Run List is [recipe[hello]]
==> foo: [2014-12-16T06:33:19+00:00] INFO: Run List expands to [hello]
==> foo: [2014-12-16T06:33:19+00:00] INFO: Starting Chef Run for vagrant.vm
==> foo: [2014-12-16T06:33:19+00:00] INFO: Running start handlers
==> foo: [2014-12-16T06:33:19+00:00] INFO: Start handlers complete.
==> foo: [2014-12-16T06:33:19+00:00] INFO: Chef Run complete in 0.0128905 seconds
==> foo: [2014-12-16T06:33:19+00:00] INFO: Running report handlers
==> foo: [2014-12-16T06:33:19+00:00] INFO: Report handlers complete
==> foo: Configuring cache buckets...
```

Checking the "/tmp/vagrant-chef-*/" directories confirms that:
```
D:\Repos\_github\bills-kitchen\target\build\repo\tmp>vagrant ssh -c "ls -la /tmp/vagrant-chef-*/"
total 20
drwxr-xr-x 3 vagrant root    4096 Dec 16 06:33 .
drwxrwxrwt 4 root    root    4096 Dec 16 06:33 ..
drwxr-xr-x 3 root    root    4096 Dec 16 06:32 chef-solo-1
-rw-rw-r-- 1 vagrant vagrant   47 Dec 16 06:33 dna.json
-rw-rw-r-- 1 vagrant vagrant  437 Dec 16 06:33 solo.rb
Connection to 127.0.0.1 closed.
```

Now, running `vagrant provision` I get the error mentioned above:
```
D:\Repos\_github\bills-kitchen\target\build\repo\tmp>vagrant provision
==> foo: Chef 11.16.4 Omnibus package is already installed.
==> foo: Configuring cache buckets...
==> foo: Running provisioner: chef_solo...
==> foo: Detected Chef (latest) is already installed
Shared folders that Chef requires are missing on the virtual machine.
This is usually due to configuration changing after already booting the
machine. The fix is to run a `vagrant reload` so that the proper shared
folders will be prepared and mounted on the VM.
```

Looking at the "/tmp/vagrant-chef-*/" directories shows that the empty "/tmp/vagrant-chef-2/" directory had been created, but the cookbooks are still under "/tmp/vagrant-chef-3/":
```
D:\Repos\_github\bills-kitchen\target\build\repo\tmp>vagrant ssh -c "ls -la /tmp/vagrant-chef-*/"
/tmp/vagrant-chef-2/:
total 8
drwxr-xr-x 2 vagrant root 4096 Dec 16 06:37 .
drwxrwxrwt 5 root    root 4096 Dec 16 06:37 ..

/tmp/vagrant-chef-3/:
total 20
drwxr-xr-x 3 vagrant root    4096 Dec 16 06:33 .
drwxrwxrwt 5 root    root    4096 Dec 16 06:37 ..
drwxr-xr-x 3 root    root    4096 Dec 16 06:32 chef-solo-1
-rw-rw-r-- 1 vagrant vagrant   47 Dec 16 06:33 dna.json
-rw-rw-r-- 1 vagrant vagrant  437 Dec 16 06:33 solo.rb
Connection to 127.0.0.1 closed.
```

The provisioner looks for the first one ("/tmp/vagrant-chef-2/chef-solo-1/"), and that's why it fails.

This happened on a Win7 64 machine with Vagrant 1.7.1. 

Any ideas?

P.S.: I still have the vagrant-omnibus plugin installed as it supports caching via vagrant-cachier, which I have not tried with the chef auto-installer feature yet.

**Comment by mitchellh** [comment]
I reproduced this as well as @sethvargo. We're taking a look.

**Comment by mitchellh** [comment]
More debug info: Here are the callstacks of when the counter is incremented for the folder that is wrong: https://gist.github.com/mitchellh/98d16d942ee77ef37e9d

**Comment by mitchellh** [comment]
Figured out why this was introduced in 1.7: The default provider logic (`Environment#default_provider`) has to load the configuration. It only has to do this on the first `up` (from not created state), but this increases the counter count by 1 vs. `vagrant provision`, a `vagrant up` from halt, etc.

Not sure how to fix this yet, but that is 100% the cause.

**Comment by mitchellh** [comment]
I think the proper long term solution is to get rid of that update counter thing and use the synced folder cache we introduced in Vagrant 1.5 to read the synced folders. That is a significant chunk of work, though.

In the short term, the best approach is probably to make the default deterministic. **BUT** I can't think of a way to sanely do this. 

I'll take a stab at the synced folder cache thing and see where I can get.

History lesson: We have all this `get_and_update_counter` stuff since basically Vagrant 0.3 or 0.4 because ever since we supported multiple provisioners we needed a way for them to not collide. In fact, this system is pretty broken. If you remove one Chef block and replace it with another, then `vagrant provision` will work but with completely the wrong Chef (or Puppet, which uses the same system). 

In Vagrant 1.5 we introduced the "synced folder cache" mechanism, where synced folders are saved to a local JSON file in the machine dir (".vagrant") on every sync folder sync. This can be queried to check what folders _should_ exist. 

Since we have that system now, we should be able to just dump all this counter stuff, use a random string for the folder name, and then look up the cache to figure it out.
